servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/ChannelSet.java

```diff
@@ -146,29 +146,47 @@ protected void handleSubscribe(final Subscriber subscriber) {
-                        // Hence, we override the offloading on each channel for this particular subscribe to use
-                        // immediate and effectively disable offloading on each channel.
-                        closeable.merge(new AsyncCloseable() {
-                            @Override
-                            public Completable closeAsync() {
-                                return channelCloseable.closeAsync().publishOnOverride(immediate());
-                            }
-
-                            @Override
-                            public Completable closeAsyncGracefully() {
-                                return channelCloseable.closeAsyncGracefully().publishOnOverride(immediate());
-                            }
-                        });
+                    if (null != channelCloseable) {
+                        // Upon shutdown of the set, we will close all live channels. If close of individual
+                        // channels are offloaded, then this would trigger a surge in threads required to offload
+                        // these closures.
+                        if (channelCloseable instanceof NettyChannelListenableAsyncCloseable) {
+                            NettyChannelListenableAsyncCloseable asNCLAC =
+                                    (NettyChannelListenableAsyncCloseable) channelCloseable;
+                            // Here we disable offloading on each channel.
+                            closeable.merge(new AsyncCloseable() {
+                                @Override
+                                public Completable closeAsync() {
+                                    return asNCLAC.closeAsyncNoOffload();
+                                }
+
+                                @Override
+                                public Completable closeAsyncGracefully() {
+                                    return asNCLAC.closeAsyncGracefullyNoOffload();
+                                }
+                            });
+                        } else {
+                            // Here we assume that if there is any offloading required, it is done by offloading the
+                            // Completable returned by closeAsyncGracefully() hence offloading each channel is not
+                            // required. Hence, we override the offloading on each channel for this particular
+                            // subscribe to use immediate() and hopefully disable offloading on each channel. There
+                            // could still be offloading embedded within the Completable but we can't override that.
+                            closeable.merge(new AsyncCloseable() {
+                                @Override
+                                public Completable closeAsync() {
+                                    return channelCloseable.closeAsync().publishOn(immediate());
```

Comment:
Does it make sense to do `publishOn(immediate())` if `closeAsync()` is already offloaded? Let's say `closeAsync()` is completed on thread A. IIUC, `publishOn(immediate())` won't change anything and will continue on thread A.